Userland/DevTools/HackStudio/HackStudioWidget.cpp

```diff
@@ -633,70 +648,153 @@ NonnullRefPtr<GUI::Action> HackStudioWidget::create_new_project_action()
-        wrapper->editor().set_editing_engine(make<GUI::RegularEditingEngine>());
+        wrapper.editor().set_editing_engine(make<GUI::RegularEditingEngine>());
     else if (previous_editor_wrapper && previous_editor_wrapper->editor().editing_engine()->is_vim())
-        wrapper->editor().set_editing_engine(make<GUI::VimEditingEngine>());
+        wrapper.editor().set_editing_engine(make<GUI::VimEditingEngine>());
     else
-        wrapper->editor().set_editing_engine(make<GUI::RegularEditingEngine>());
+        wrapper.editor().set_editing_engine(make<GUI::RegularEditingEngine>());
+
+    wrapper.on_tab_close_request = [this, &parent](auto& tab) {
+        parent.deferred_invoke([this, &parent, &tab] {
+            parent.remove_tab(tab);
+            m_all_editor_wrappers.remove_first_matching([&tab](auto& entry) { return entry == &tab; });
+            if (parent.children().is_empty()) {
+                m_switch_to_next_editor_tab_widget->activate();
+                m_editors_splitter->remove_child(parent);
+                m_all_editor_tab_widgets.remove_first_matching([&parent](auto& entry) { return entry == &parent; });
+                if (m_all_editor_tab_widgets.size() > 1) {
+                    for (auto& widget : m_all_editor_tab_widgets)
+                        widget.set_close_button_enabled(true);
+                } else {
+                    m_all_editor_tab_widgets[0].set_close_button_enabled(false);
+                }
+            }
+            if (parent.children().size() == 1 && m_all_editor_tab_widgets.size() <= 1)
+                parent.set_close_button_enabled(false);
+            update_actions();
+        });
+    };
 }
 
-NonnullRefPtr<GUI::Action> HackStudioWidget::create_switch_to_next_editor_action()
+NonnullRefPtr<GUI::Action> HackStudioWidget::create_switch_to_next_editor_tab_widget_action()
 {
-    return GUI::Action::create("Switch to &Next Editor", { Mod_Ctrl, Key_E }, [this](auto&) {
-        if (m_all_editor_wrappers.size() <= 1)
+    return GUI::Action::create("Switch to Next Editor Group", { Mod_Ctrl | Mod_Shift, Key_T }, [this](auto&) {
+        if (m_all_editor_tab_widgets.size() <= 1)
             return;
-        Vector<EditorWrapper&> wrappers;
-        m_editors_splitter->for_each_child_of_type<EditorWrapper>([&wrappers](auto& child) {
-            wrappers.append(child);
+        Vector<GUI::TabWidget&> tab_widgets;
+        m_editors_splitter->for_each_child_of_type<GUI::TabWidget>([&tab_widgets](auto& child) {
+            tab_widgets.append(child);
             return IterationDecision::Continue;
         });
-        for (size_t i = 0; i < wrappers.size(); ++i) {
-            if (m_current_editor_wrapper.ptr() == &wrappers[i]) {
-                if (i == wrappers.size() - 1)
-                    wrappers[0].editor().set_focus(true);
+        for (size_t i = 0; i < tab_widgets.size(); ++i) {
+            if (m_current_editor_tab_widget.ptr() == &tab_widgets[i]) {
```

Comment:
Switching tab widgets doesn't work when the current tab widget is at the very top. I need to spend some time on this logic.